src/middleware.rs

```diff
@@ -164,3 +164,87 @@ where
         (self)(request, next)
     }
 }
+
+#[cfg(feature = "digest-auth")]
+pub mod digest {
+    use super::*;
+    use crate::{Request, Response};
+    use digest_auth::{AuthContext, WwwAuthenticateHeader};
+    use std::{borrow::Cow, str::FromStr};
+
+    /// Provides simple digest authentication powered by the `digest_auth` crate.
+    ///
+    /// Requests that receive a HTTP 401 response are retried once by this middleware with the
+    /// credentials provided on construction. The retry only happens under these conditions:
+    /// - there is no prior "authorization" header on the request set by the caller or other
+    ///   middleware, and;
+    /// - the server provides HTTP Digest auth challenge in the "www-authenticate" header.
+    ///
+    /// In other cases, this middleware acts as a no-op forwarder of requests and responses.
+    ///
+    /// ```
+    /// let arbitrary_username = "MyUsername";
+    /// let arbitrary_password = "MyPassword";
+    /// let digest_auth_middleware =
+    ///     ureq::DigestAuthMiddleware::new(arbitrary_username, arbitrary_password);
+    /// # let url = String::new();
+    ///
+    /// let agent = ureq::AgentBuilder::new().middleware(digest_auth_middleware).build();
+    /// agent.get(&url).call();
+    /// ```
+    pub struct DigestAuthMiddleware {
+        username: Cow<'static, str>,
+        password: Cow<'static, str>,
+    }
+
+    impl DigestAuthMiddleware {
+        pub fn new(
+            username: impl Into<Cow<'static, str>>,
+            password: impl Into<Cow<'static, str>>,
```

Comment:
We have an informal policy of avoiding "complicated" type signatures, like those involving cows 🐮. In user facing PI, we prefer these to be just `&str`, even though it's a bit less efficient. It is however fine to have cows internally.

Reply:
I'm not sure I can make Cows too useful with a &str input type as I need to satisfy the `'static` bound, but I'll replace Cows with Strings since this all seems like premature optimization anyway :)